feat: Serverless VPC Access Connector update fields without recreating the resource

ramonvermeulen opened the pull request:
<!--
Complete the self-review checklist to help speed up the review process: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

If your PR is still work in progress, please create it in draft mode.

Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to.
For example: Fixes https://github.com/hashicorp/terraform-provider-google/issues/ISSUE_ID
-->
Solves https://github.com/hashicorp/terraform-provider-google/issues/20814

Added functionality so that updates on the fields [`min_instances`, `max_instances`, `machine_type`] for the `google_vpc_access_connector` resource will not cause recreation, and use the underlaying [**projects.locations.connectors/patch**](https://cloud.google.com/vpc/docs/reference/vpcaccess/rest/v1/projects.locations.connectors/patch) endpoint instead to update in-place.

**Release Note Template for Downstream PRs (will be copied)**

See [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) for guidance.

```release-note:enhancement
vpcaccess: changed fields `min_instances`, `max_instances`, `machine_type` to allow update `google_vpc_access_connector` without without recreation.
```

ramonvermeulen commented:
Unrelated question to the PR:

Is there a good way to find out which fields are mutable/immutable on the GCP API Endpoints? As far I can see this is not something that the endpoint documentation provides (or I am blind). I had to find out via a combination of checking the GCP Console UI, but then I realized that maybe some fields could be supported via API only. Then I saw this message when running the tests the first time:

```Error: Error updating Connector "projects/XXXXX/locations/us-central1/connectors/tf-test-mqice1t7dx": googleapi: Error 400: Update mask contains illegal fields. Allowed fields are: [min_instances, max_instances, machine_type]```

And then I figured these were the fields, probably there is a simpler way.

rileykarson commented:
I haven't found one! Those match the steps I'd take personally.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 197 insertions(+), 25 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 197 insertions(+), 25 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 7
Passed tests: 3
Skipped tests: 1
Affected tests: 3

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>vpcaccess</li>

</ul>

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 3 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccVPCAccessConnectorDatasource_basic</li>
<li>TestAccVPCAccessConnectorDatasource_update</li>
<li>TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🟢 Tests passed during RECORDING mode:
`TestAccVPCAccessConnectorDatasource_update` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/86396d5e-3ede-4fab-8bc0-3cd0047b7fc4/recording/TestAccVPCAccessConnectorDatasource_update.log)]

🟢 No issues found for passed tests after REPLAYING rerun.

---


🔴 Tests failed during RECORDING mode:
`TestAccVPCAccessConnectorDatasource_basic` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/86396d5e-3ede-4fab-8bc0-3cd0047b7fc4/build-log/recording_build/TestAccVPCAccessConnectorDatasource_basic_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/86396d5e-3ede-4fab-8bc0-3cd0047b7fc4/recording/TestAccVPCAccessConnectorDatasource_basic.log)]
`TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/86396d5e-3ede-4fab-8bc0-3cd0047b7fc4/build-log/recording_build/TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/86396d5e-3ede-4fab-8bc0-3cd0047b7fc4/recording/TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits.log)]



🔴 Errors occurred during RECORDING mode. Please fix them to complete your PR.



View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/86396d5e-3ede-4fab-8bc0-3cd0047b7fc4/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/86396d5e-3ede-4fab-8bc0-3cd0047b7fc4/recording) for each test

rileykarson reviewed (commented):
Comments inline. Re: failed tests:

`TestAccVPCAccessConnectorDatasource_basic` - `Error: Error waiting to create Connector: Error waiting for Creating Connector: Error code 3, message: Operation failed: Invalid IP CIDR range was provided. It conflicts with an existing subnetwork. Please delete the connector manually.` so we'll need to pick a diff. CIDR.

`TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits` was doing a recreation internally before this change and attempts to update now- that's a bit of a weird test and should be split up, but this change does interact unusually with the min_/max_ instance and throughput fields and can create invalid updates. Let me see if I can think of a good mitigation.

ramonvermeulen commented on mmv1/third_party/terraform/services/vpcaccess/data_source_vpc_access_connector_test.go:
```diff
@@ -1,7 +1,7 @@
 package vpcaccess_test
```
Thanks for the review! I just spotted it as well when looking at `resource_vpc_access_connector_test.go`, sorry it is my first time working with magic_modules so still learning a lot ;)

I see the tests in that file make a lot more sense, and also has a nice example of testing an Error, so I will revert the changes in `data_source_vpc_access_connector_test.go` and write two tests in `resource_vpc_access_connector_test.go`:
1. happy scenario, just changing `[min_instances, max_instances, machine_type]` expecting an update.
2. un-happy scenario, trying to change an immutable property, expecting an error.

ramonvermeulen commented:
> Comments inline. Re: failed tests:
> 
> `TestAccVPCAccessConnectorDatasource_basic` - `Error: Error waiting to create Connector: Error waiting for Creating Connector: Error code 3, message: Operation failed: Invalid IP CIDR range was provided. It conflicts with an existing subnetwork. Please delete the connector manually.` so we'll need to pick a diff. CIDR.
> 
> `TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits` was doing a recreation internally before this change and attempts to update now- that's a bit of a weird test and should be split up, but this change does interact unusually with the min_/max_ instance and throughput fields and can create invalid updates. Let me see if I can think of a good mitigation.

Thanks for your review! @rileykarson 

Just reverted the datasource changes, I see the original test was already using `10.8.0.32/28` so let's see if this causes a problem again for `TestAccVPCAccessConnectorDatasource_basic`, otherwise I will change it to another CIDR range.

In `resource_vpc_access_connector_test.go` I implemented two tests, one expecting an in-place update when changing all the mutable fields. Another one triggering a replacement when changing an immutable field.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 245 insertions(+), 3 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 245 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 8
Passed tests: 4
Skipped tests: 1
Affected tests: 3

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>vpcaccess</li>

</ul>

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 3 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits</li>
<li>TestAccVPCAccessConnector_vpcAccessConnectorUpdateAllMutableFieldsCauseUpdateInPlace</li>
<li>TestAccVPCAccessConnector_vpcAccessConnectorUpdateImmutableFieldCauseReplace</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🟢 Tests passed during RECORDING mode:
`TestAccVPCAccessConnector_vpcAccessConnectorUpdateAllMutableFieldsCauseUpdateInPlace` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/d98d02bf-3939-4eae-b0d8-e7daf77d2a46/recording/TestAccVPCAccessConnector_vpcAccessConnectorUpdateAllMutableFieldsCauseUpdateInPlace.log)]
`TestAccVPCAccessConnector_vpcAccessConnectorUpdateImmutableFieldCauseReplace` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/d98d02bf-3939-4eae-b0d8-e7daf77d2a46/recording/TestAccVPCAccessConnector_vpcAccessConnectorUpdateImmutableFieldCauseReplace.log)]

🟢 No issues found for passed tests after REPLAYING rerun.

---


🔴 Tests failed during RECORDING mode:
`TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/d98d02bf-3939-4eae-b0d8-e7daf77d2a46/build-log/recording_build/TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/d98d02bf-3939-4eae-b0d8-e7daf77d2a46/recording/TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits.log)]



🔴 Errors occurred during RECORDING mode. Please fix them to complete your PR.



View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/d98d02bf-3939-4eae-b0d8-e7daf77d2a46/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/d98d02bf-3939-4eae-b0d8-e7daf77d2a46/recording) for each test

ramonvermeulen commented:
@rileykarson anything I can do to speed-up the process? Unfortunately I don't have access to see the test outputs, so it's hard to determine if `TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits` is related to the changes I've made.

rileykarson commented:
Let me rerun to see if that's now stale- the error is "Error: Error updating Connector "projects/ci-test-project-188019/locations/us-central1/connectors/tf-test-vpc-cone4bex9pmmm": googleapi: Error 400: The maximum number of instances cannot be set below the current value." which indicates we are sending a bad request message, I think sending a stale value for one of the mutually exclusive sizing fields.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 245 insertions(+), 3 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 245 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 8
Passed tests: 6
Skipped tests: 1
Affected tests: 1

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>vpcaccess</li>

</ul>

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🔴 Tests failed during RECORDING mode:
`TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/36196169-2f69-42e3-8b57-aacf0e5ffa5c/build-log/recording_build/TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/36196169-2f69-42e3-8b57-aacf0e5ffa5c/recording/TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits.log)]



🔴 Errors occurred during RECORDING mode. Please fix them to complete your PR.



View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/36196169-2f69-42e3-8b57-aacf0e5ffa5c/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/36196169-2f69-42e3-8b57-aacf0e5ffa5c/recording) for each test

ramonvermeulen commented:
> Let me rerun to see if that's now stale- the error is "Error: Error updating Connector "projects/ci-test-project-188019/locations/us-central1/connectors/tf-test-vpc-cone4bex9pmmm": googleapi: Error 400: The maximum number of instances cannot be set below the current value." which indicates we are sending a bad request message, I think sending a stale value for one of the mutually exclusive sizing fields.

@rileykarson Thanks for sharing the error message! Seems like this error occurs in this second test case, I assume on this line:
https://github.com/GoogleCloudPlatform/magic-modules/pull/12830/files#diff-c3434d2bb5e0b8b19cf03b41a500e24b0845707d01693c90f3440dd75962685eR181

The first test case sets the min_throughput to `400` and max_throughput to `800` to test the behavior that that should also automatically set the max_instances to `4` and `8`. Then the second test case sets the max_instances to `7` and min_instances to `5` on the same resource. Might it be that this is something that has changed in the API, since I don't see a relation to the changes I've made? I assume the error is about changing this max_instances value from `8` to `7`.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 245 insertions(+), 3 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 245 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 8
Passed tests: 6
Skipped tests: 1
Affected tests: 1

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>vpcaccess</li>

</ul>

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🔴 Tests failed during RECORDING mode:
`TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/c7371d9c-e113-4f9b-aa66-5faf09a42ae4/build-log/recording_build/TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/c7371d9c-e113-4f9b-aa66-5faf09a42ae4/recording/TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits.log)]



🔴 Errors occurred during RECORDING mode. Please fix them to complete your PR.



View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/c7371d9c-e113-4f9b-aa66-5faf09a42ae4/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/c7371d9c-e113-4f9b-aa66-5faf09a42ae4/recording) for each test

roaks3 commented:
@modular-magician reassign-reviewer

ScottSuarez commented on mmv1/products/vpcaccess/Connector.yaml:
```diff
@@ -22,7 +22,10 @@ references:
+update_url: 'projects/{{project}}/locations/{{region}}/connectors/{{name}}'
+update_verb: 'PATCH'
+update_mask: true
+immutable: false
```
```suggestion
```
implied by absence

ScottSuarez reviewed (commented):
small ask

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 245 insertions(+), 3 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 245 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 8
Passed tests: 6
Skipped tests: 1
Affected tests: 1

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>vpcaccess</li>

</ul>

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🔴 Tests failed during RECORDING mode:
`TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/059e01ce-0e65-4ebd-aceb-93539fc8a251/build-log/recording_build/TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/059e01ce-0e65-4ebd-aceb-93539fc8a251/recording/TestAccVPCAccessConnector_vpcAccessConnectorThroughput_usingThroughputOrInstancesLimits.log)]



🔴 Errors occurred during RECORDING mode. Please fix them to complete your PR.



View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/059e01ce-0e65-4ebd-aceb-93539fc8a251/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/059e01ce-0e65-4ebd-aceb-93539fc8a251/recording) for each test

ScottSuarez reviewed (changes requested):
Actually upon review I believe this is breaking testcase
`testAccVPCAccessConnector_vpcAccessConnectorThroughput_justInstanceFields`.

Since these fields were not previously updatable, allowing them to be so is giving us an api error for these fields. Cannot grow/shrink on certain fields.

To maintain current behavior we will need to implement a custom diff

```
            customdiff.ForceNewIfChange("size", func (ctx context.Context, old, new, meta any) bool {
                // "size" can only increase in-place, so we must create a new resource
                // if it is decreased.
                return new.(int) < old.(int)
            }),
```

See  
https://github.com/ScottSuarez/magic-modules/blob/8003330c2e4819d3488bc77157128f1cf4cd1369/mmv1/products/apphub/Application.yaml#L43

https://github.com/GoogleCloudPlatform/magic-modules/blob/3a77e6cecf22162ff745b212752b67d528206bed/tpgtools/templates/resource.go.tmpl#L80-L86

https://github.com/GoogleCloudPlatform/magic-modules/blob/3a77e6cecf22162ff745b212752b67d528206bed/mmv1/api/resource.go#L208

ramonvermeulen commented:
> ustomdiff.ForceNewIfChange("size",
Thanks for sharing this @ScottSuarez 

I did some tests, and you are indeed right. `min_instances` and `max_instances` **can only be updated in place when being increased**, otherwise it requires recreation. Will look a bit into how this custom_diff works, to see if I can implement it correctly. 

E.g.: desired behavior here is, when you decrease the `min_instances` or `max_instances` the resource should be recreated instead of updated in-place.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 257 insertions(+), 3 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 257 insertions(+), 3 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12830-old..auto-pr-12830) ( 1 file changed, 10 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: 8
Passed tests: 7
Skipped tests: 1
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>vpcaccess</li>

</ul>

</blockquote>
</details>

🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/28cbba10-7bf5-466e-baab-595461f5ae8b/build-log/replaying_test.log)

ScottSuarez reviewed (changes requested):
So we are getting closer.

When testing locally, I was unable to change one value without changing the other. Can you enable it so the min value can change without changing max value. To do this you will need to write a `custom_code`, `pre_update` to set the max or mix on the api obj the other is set.

Also, min and max should be required_with each other
https://github.com/ScottSuarez/magic-modules/blob/8003330c2e4819d3488bc77157128f1cf4cd1369/mmv1/api/type.go#L135

ramonvermeulen commented:
> So we are getting closer.
> 
> When testing locally, I was unable to change one value without changing the other. Can you enable it so the min value can change without changing max value. To do this you will need to write a `custom_code`, `pre_update` to set the max or mix on the api obj if the other is set.
> 
> Also, min and max should be required_with each other https://github.com/ScottSuarez/magic-modules/blob/8003330c2e4819d3488bc77157128f1cf4cd1369/mmv1/api/type.go#L135

@ScottSuarez 

Thanks again for your review! I tried to implement it correctly based on your information, learning a lot about magic_modules here :)

Currently still struggling a bit to find a nice local testing set-up, I am building the provider via:

```
make provider VERSION=ga OUTPUT_PATH="$GOPATH/src/github.com/hashicorp/terraform-provider-google" PRODUCT=vpcaccess
```

However, I am looking for a quick way to then use my own "local" build of the provider with a simple terraform file to test the specific resource and play around with the arguments.

ScottSuarez reviewed (changes requested):
Can you add a test for min changing without max changing?

rileykarson commented:
Doing a flow like that is documented at https://googlecloudplatform.github.io/magic-modules/test/run-tests/#optional-test-manually fwiw

ramonvermeulen commented:
> Doing a flow like that is documented at https://googlecloudplatform.github.io/magic-modules/test/run-tests/#optional-test-manually fwiw

My bad, thanks! I think I overlooked this part in the documentation. 

Thanks (to both of you) for all the information, I will add a test for this scenario where only one of the two attributes changes, and probably finalize my PR by the end of the week.

ramonvermeulen commented:
I am getting closer, at least I implemented a test for this scenario now (that fails).

Weirdly the error message seems weird, when I run with TF_LOG=DEBUG it seems to provide both `minInstances` as `maxInstances` but still gives an error.

![example2](https://github.com/user-attachments/assets/64ae762f-ab7b-4659-99a1-75b80d176a29)

Will debug further to find the cause.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 347 insertions(+), 3 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 347 insertions(+), 3 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12830-old..auto-pr-12830) ( 1 file changed, 10 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: 9
Passed tests: 4
Skipped tests: 1
Affected tests: 4

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>vpcaccess</li>

</ul>

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 4 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccVPCAccessConnector_vpcAccessConnectorThroughput</li>
<li>TestAccVPCAccessConnector_vpcAccessConnectorUpdateAllMutableFieldsCauseUpdateInPlace</li>
<li>TestAccVPCAccessConnector_vpcAccessConnectorUpdateImmutableFieldCauseReplace</li>
<li>TestAccVPCAccessConnector_vpcAccessConnectorUpdateOnlyMinInstancesCauseUpdateInPlace</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🟢 Tests passed during RECORDING mode:
`TestAccVPCAccessConnector_vpcAccessConnectorThroughput` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/c14c9d55-dfef-4af0-87d7-7e844232b1be/recording/TestAccVPCAccessConnector_vpcAccessConnectorThroughput.log)]
`TestAccVPCAccessConnector_vpcAccessConnectorUpdateAllMutableFieldsCauseUpdateInPlace` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/c14c9d55-dfef-4af0-87d7-7e844232b1be/recording/TestAccVPCAccessConnector_vpcAccessConnectorUpdateAllMutableFieldsCauseUpdateInPlace.log)]
`TestAccVPCAccessConnector_vpcAccessConnectorUpdateImmutableFieldCauseReplace` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/c14c9d55-dfef-4af0-87d7-7e844232b1be/recording/TestAccVPCAccessConnector_vpcAccessConnectorUpdateImmutableFieldCauseReplace.log)]
`TestAccVPCAccessConnector_vpcAccessConnectorUpdateOnlyMinInstancesCauseUpdateInPlace` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/c14c9d55-dfef-4af0-87d7-7e844232b1be/recording/TestAccVPCAccessConnector_vpcAccessConnectorUpdateOnlyMinInstancesCauseUpdateInPlace.log)]

🟢 No issues found for passed tests after REPLAYING rerun.

---






🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/c14c9d55-dfef-4af0-87d7-7e844232b1be/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/c14c9d55-dfef-4af0-87d7-7e844232b1be/recording) for each test

ramonvermeulen commented on mmv1/templates/terraform/pre_update/vpc_access_connector_instances.go.tmpl:
```diff
@@ -0,0 +1,19 @@
+}
+
+if d.HasChange("machine_type") {
+    newUpdateMask = append(updateMask, "machineType")
```
@ScottSuarez 

Because we are fully overriding the "updateMask" via `url, err = transport_tpg.AddQueryParams(url, map[string]string{"updateMask": strings.Join(newUpdateMask, ",")})` I thought that it is necessary to implement this for all members (maybe there is a way to append instead, not sure if the existing updateMask is accessible within this template scope)?

Otherwise I assume updateMask will be empty if you purely update `machineType`, these 3 fields are the only fields that allow in-place updates. Rest will do a recreate of the resource anyways.

ScottSuarez replied:
If you look at the generated code
https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12830-old..auto-pr-12830#diff-3e57c758e5b7a09f15b99d9099442e0d5581c25a29391e9fc42c3bb2061a3b0aR489-R524

The original update mask is in scope. Lets reuse that here. If we were to maintain a new update mask field that overrides the other one then fields in the future that support update would manually be maintained.

To understand how/where your code is injected you can take a look at the generated files over at:
https://github.com/ScottSuarez/magic-modules/blob/7fa80658c0ead5195bfb361d52a2fa15396c7c28/mmv1/templates/terraform/resource.go.tmpl#L784

ScottSuarez commented on mmv1/templates/terraform/pre_update/vpc_access_connector_instances.go.tmpl:
```diff
@@ -0,0 +1,21 @@
+newUpdateMask := []string{}
+
+if d.HasChange("min_instances") && !d.HasChange("max_instances") {
+    obj["maxInstances"] = d.Get("max_instances").(int)
+    newUpdateMask = append(newUpdateMask, "maxInstances", "minInstances")
+}
+
+if d.HasChange("max_instances") && !d.HasChange("min_instances") {
+    obj["minInstances"] = d.Get("min_instances").(int)
+    newUpdateMask = append(newUpdateMask, "maxInstances", "minInstances")
+}
+
+if d.HasChange("machine_type") {
+    newUpdateMask = append(updateMask, "machineType")
+}
+
+// Overwrite the previously set mask.
+url, err = transport_tpg.AddQueryParams(url, map[string]string{"updateMask": strings.Join(newUpdateMask, ",")})
+if err != nil {
+    return err
+}
```
```suggestion
newUpdateMask := []string{}

if d.HasChange("min_instances") && !d.HasChange("max_instances") {
    obj["maxInstances"] = d.Get("max_instances").(int)
    updateMask = append(updateMask, "maxInstances", "minInstances")
}

if d.HasChange("max_instances") && !d.HasChange("min_instances") {
    obj["minInstances"] = d.Get("min_instances").(int)
    updateMask = append(updateMask, "maxInstances", "minInstances")
}

// Overwrite the previously set mask.
url, err = transport_tpg.AddQueryParams(url, map[string]string{"updateMask": strings.Join(updateMask, ",")})
if err != nil {
    return err
}
```

ramonvermeulen replied:
Thanks! Looking at the generated code makes things indeed a lot more clear :)

I totally get your point about maintainability, so applied your changes to re-use the existing `updateMask` in scope.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 343 insertions(+), 3 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12830-old..auto-pr-12830) ( 3 files changed, 343 insertions(+), 3 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12830-old..auto-pr-12830) ( 1 file changed, 10 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: 9
Passed tests: 6
Skipped tests: 1
Affected tests: 2

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>vpcaccess</li>

</ul>

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 2 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccVPCAccessConnector_vpcAccessConnectorUpdateAllMutableFieldsCauseUpdateInPlace</li>
<li>TestAccVPCAccessConnector_vpcAccessConnectorUpdateOnlyMinInstancesCauseUpdateInPlace</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🟢 Tests passed during RECORDING mode:
`TestAccVPCAccessConnector_vpcAccessConnectorUpdateAllMutableFieldsCauseUpdateInPlace` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/78e70efd-e122-49bf-a0c0-a635022bd668/recording/TestAccVPCAccessConnector_vpcAccessConnectorUpdateAllMutableFieldsCauseUpdateInPlace.log)]
`TestAccVPCAccessConnector_vpcAccessConnectorUpdateOnlyMinInstancesCauseUpdateInPlace` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/78e70efd-e122-49bf-a0c0-a635022bd668/recording/TestAccVPCAccessConnector_vpcAccessConnectorUpdateOnlyMinInstancesCauseUpdateInPlace.log)]

🟢 No issues found for passed tests after REPLAYING rerun.

---






🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/78e70efd-e122-49bf-a0c0-a635022bd668/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-12830/artifacts/78e70efd-e122-49bf-a0c0-a635022bd668/recording) for each test

ScottSuarez commented:
Looks good ! Thanks for the work and persistence here!

ramonvermeulen commented:
@ScottSuarez thanks for the support on this one, learned a lot! 👌
